src/django_pg_migration_tools/operations.py

```diff
@@ -525,3 +583,256 @@ def describe(self) -> str:
             f"using a UNIQUE index. NOTE: Using the django_pg_migration_tools "
             f"SaferRemoveIndexConcurrently operation."
         )
+
+
+class NullsManager(base_operations.Operation):
+    def set_not_null(
+        self,
+        app_label: str,
+        schema_editor: base_schema.BaseDatabaseSchemaEditor,
+        model: type[models.Model],
+        column_name: str,
+    ) -> None:
+        """
+        The following routine takes into consideration:
+
+        Reentrancy:
+          Any of the given operations below may fail for a variety of
+          reasons. If a failure occurs, this routine can be rerun without
+          impacting correct execution. Each step progresses toward the
+          outcome while leaving the db in a consistent state.
+          If the operation fails mid-way through, it can still be picked up
+          again. The database isn't left in an inconsistent state.
+
+        Idempotency:
+          The code expects clients to retry migrations without extra
+          side-effects. So does this routine. Multiple calls to the
+          routine have the same effect on the system state as a single
+          call.
+
+        Small number of introspective SQL queries:
+          Introspective SQL queries are necessary for checking the state of
+          the database. This is required for idempotency and reentrancy. At
+          each step, the routine only fires as few introspective SQL
+          statements as necessary.
+        """
+        psql_operations.NotInTransactionMixin()._ensure_not_in_transaction(
+            schema_editor
+        )
+        if not self.allow_migrate_model(schema_editor.connection.alias, model):
+            return
+
+        table_name = model._meta.db_table
+        constraint_name = self._get_constraint_name(table_name, column_name)
+
+        # Just checking for NOT NULL isn't sufficient, we need to check that
+        # the constraint doesn't exist (VALID or NOT). Even a NOT VALID
+        # constraint would set the field to be non-nullable.
+        is_not_null = self._is_not_null(schema_editor, table_name, column_name)
+        constraint_exists = self._constraint_exists(schema_editor, constraint_name)
+        if is_not_null and (not constraint_exists):
+            return
```

Comment:
Why are we doing these checks? (It's to enable reentrancy and idempotency.) The explanations here don't make much sense without that context.

Reply:
My original intend was to explain that a NOT VALID constraint would set the
catalogue entry for this table attr to be NOT NULL, so querying for the
catalogue is not sufficient. I think this comment shouldn't be here anymore.
e2b4075df2c6de380a28426dd1875eb9e088018d